services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -3057,25 +2760,71 @@ private SchemaEntry addValueSchemaEntry(
+  public void addSupersetSchema(
       String clusterName,
       String storeName,
       String valueSchemaStr,
       int valueSchemaId,
       String supersetSchemaStr,
       int supersetSchemaId) {
-    throw new VeniceUnsupportedOperationException("addSupersetSchema");
+    acquireAdminMessageLock(clusterName, storeName);
+    try {
+      if (supersetSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+        throw new VeniceException("Invalid superset schema id: " + supersetSchemaId);
+      }
+
+      ReadWriteSchemaRepository schemaRepository = getHelixVeniceClusterResources(clusterName).getSchemaRepository();
+      final SchemaEntry existingSupersetSchemaEntry = schemaRepository.getValueSchema(storeName, supersetSchemaId);
+      if (existingSupersetSchemaEntry != null) {
+        final Schema newSupersetSchema = AvroSchemaParseUtils.parseSchemaFromJSONStrictValidation(supersetSchemaStr);
+        if (!AvroSchemaUtils
+            .compareSchemaIgnoreFieldOrder(existingSupersetSchemaEntry.getSchema(), newSupersetSchema)) {
+          throw new VeniceException(
+              "Existing schema with id " + existingSupersetSchemaEntry.getId() + " does not match with new schema "
+                  + supersetSchemaStr);
+        }
+      }
+
+      if (valueSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+        LOGGER.info(
+            "Adding superset schema {} with id {} to store: {} in cluster: {}",
+            supersetSchemaStr,
+            supersetSchemaId,
+            storeName,
+            clusterName);
+        valueSchemaStr = "";
+      } else if (StringUtils.isEmpty(valueSchemaStr)) {
+        throw new VeniceException("Invalid value schema string: " + valueSchemaStr);
+      }
+
+      SupersetSchemaCreation supersetSchemaCreation =
+          (SupersetSchemaCreation) AdminMessageType.SUPERSET_SCHEMA_CREATION.getNewInstance();
+      supersetSchemaCreation.clusterName = clusterName;
+      supersetSchemaCreation.storeName = storeName;
+      SchemaMeta valueSchemaMeta = new SchemaMeta();
+      valueSchemaMeta.definition = valueSchemaStr;
+      valueSchemaMeta.schemaType = SchemaType.AVRO_1_4.getValue();
+      supersetSchemaCreation.valueSchema = valueSchemaMeta;
+      supersetSchemaCreation.valueSchemaId = valueSchemaId;
+
+      SchemaMeta supersetSchemaMeta = new SchemaMeta();
+      supersetSchemaMeta.definition = supersetSchemaStr;
+      supersetSchemaMeta.schemaType = SchemaType.AVRO_1_4.getValue();
+      supersetSchemaCreation.supersetSchema = supersetSchemaMeta;
+      supersetSchemaCreation.supersetSchemaId = supersetSchemaId;
+
+      AdminOperation message = new AdminOperation();
+      message.operationType = AdminMessageType.SUPERSET_SCHEMA_CREATION.getValue();
+      message.payloadUnion = supersetSchemaCreation;
+
+      sendAdminMessageAndWaitForConsumed(clusterName, storeName, message);
+    } finally {
+      releaseAdminMessageLock(clusterName, storeName);
```

Comment:
Can we guarantee that we don't run into scenario where the admin is stuck and this never executes? Is there some sort of timed lock releases?

Reply:
This is carried over from the current way we run this. Also, there is a timeout in `sendAdminMessageAndWaitForConsumed`